// djs-button-pages/dist/Constants.js
"use strict";
/**
 * Copyright © 2022 Danila Kononov (nickname: moony). All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Constants that may be changed in the future.
 */
var Constants;
(function (Constants) {
    /**
     * Current Discord's embed length limit.
     */
    Constants[Constants["DISCORD_MAX_EMBED_LENGTH"] = 6000] = "DISCORD_MAX_EMBED_LENGTH";
    /**
     * Current Discord's limit for buttons per row.
     */
    Constants[Constants["DISCORD_MAX_BUTTONS_PER_ROW"] = 5] = "DISCORD_MAX_BUTTONS_PER_ROW";
    /**
     * Current Discord's limit for rows per message.
     */
    Constants[Constants["DISCORD_MAX_ROWS_PER_MESSAGE"] = 3] = "DISCORD_MAX_ROWS_PER_MESSAGE";
    /**
     * Current Discord's limit for interaction life-time.
     */
    Constants[Constants["DISCORD_MAX_INTERACTION_LIFE_TIME"] = 900000] = "DISCORD_MAX_INTERACTION_LIFE_TIME";
    /**
     * Current library limit for max pagination life-time.
     */
    Constants[Constants["LIBRARY_MAX_PAGES_LIFE_TIME"] = 3600000] = "LIBRARY_MAX_PAGES_LIFE_TIME";
    /**
     * Current library limit for minimum pagination life-time.
     */
    Constants[Constants["LIBRARY_MIN_PAGES_LIFE_TIME"] = 1000] = "LIBRARY_MIN_PAGES_LIFE_TIME";
})(Constants || (Constants = {}));
;
exports.default = Constants;
