// djs-button-pages/dist/Paginations.js
"use strict";
/**
 * Copyright © 2022 Danila Kononov (nickname: moony). All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEmbedLength = exports.Constants = exports.ButtonData = exports.PaginationData = exports.BasePagination = exports.StopButton = exports.PreviousPageButton = exports.NextPageButton = exports.LastPageButton = exports.FirstPageButton = exports.CustomButton = exports.InteractionPagination = exports.MessageReplyPagination = exports.ChannelPagination = void 0;
const ChannelPagination_1 = require("./Classes/Paginations/ChannelPagination");
exports.ChannelPagination = ChannelPagination_1.default;
const MessageReplyPagination_1 = require("./Classes/Paginations/MessageReplyPagination");
exports.MessageReplyPagination = MessageReplyPagination_1.default;
const InteractionPagination_1 = require("./Classes/Paginations/InteractionPagination");
exports.InteractionPagination = InteractionPagination_1.default;
const CustomButton_1 = require("./Classes/Buttons/CustomButton");
exports.CustomButton = CustomButton_1.default;
const FirstPageButton_1 = require("./Classes/Buttons/FirstPageButton");
exports.FirstPageButton = FirstPageButton_1.default;
const LastPageButton_1 = require("./Classes/Buttons/LastPageButton");
exports.LastPageButton = LastPageButton_1.default;
const NextPageButton_1 = require("./Classes/Buttons/NextPageButton");
exports.NextPageButton = NextPageButton_1.default;
const PreviousPageButton_1 = require("./Classes/Buttons/PreviousPageButton");
exports.PreviousPageButton = PreviousPageButton_1.default;
const StopButton_1 = require("./Classes/Buttons/StopButton");
exports.StopButton = StopButton_1.default;
const BasePagination_1 = require("./Classes/Paginations/Basic/BasePagination");
exports.BasePagination = BasePagination_1.default;
const PaginationData_1 = require("./Classes/Paginations/Basic/PaginationData");
exports.PaginationData = PaginationData_1.default;
const ButtonData_1 = require("./Classes/Buttons/Basic/ButtonData");
exports.ButtonData = ButtonData_1.default;
const Constants_1 = require("./Constants");
exports.Constants = Constants_1.default;
const Utils_1 = require("./Utils");
Object.defineProperty(exports, "getEmbedLength", { enumerable: true, get: function () { return Utils_1.getEmbedLength; } });
