// djs-button-pages/dist/Classes/Buttons/Basic/ButtonData.js
"use strict";
/**
 * Copyright © 2022 Danila Kononov (nickname: moony). All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Class for storing and modifying button data.
 */
class ButtonData {
    /**
     * Class for storing and modifying button data.
     * @param {ButtonStyling | ButtonData} data Data from which to build ButtonData.
     */
    constructor(data) {
        if (!data)
            return;
        this._setup(data);
        return;
    }
    ;
    _style;
    _action;
    _disableWhen;
    /**
     * Style of this button.
     * @type {ButtonStyling | null}
     */
    get style() {
        return this._style ?? null;
    }
    ;
    /**
     * Represents an action that should be completed after the button is pressed.
     *
     * Either a number of a page or a function that determines number of a page.
     * @type {ButtonAction | null}
     */
    get action() {
        return this._action || this._action === 0 ? this._action : null;
    }
    ;
    /**
     * Represents a condition of disabling button.
     *
     * Either a number of a page or a function that determines number of a page.
     * @type {ButtonDisableWhen | null}
     */
    get disableWhen() {
        return this._disableWhen || this._disableWhen === 0 ? this._disableWhen : null;
    }
    ;
    /**
     * Sets button style.
     * @param {ButtonStyling} style Button style.
     * @returns {this} Button data.
     */
    setStyle(style) {
        if (!style.custom_id || (!style.emoji && !style.label) || !style.style)
            throw new Error("Button should have customId, emoji or label and style.");
        this._style = style;
        return this;
    }
    ;
    /**
     * Sets an action that should be completed when the button is pressed.
     * @param {ButtonAction} action Either a number of a page or a function that determines number of a page.
     * @returns {this} Button data.
     */
    _setAction(action) {
        if (typeof action === "number" && (!Number.isInteger(action) || action < -1))
            throw new RangeError("Action should return natural number or minus one to stop pagination.");
        this._action = action;
        return this;
    }
    ;
    /**
     * Sets a condition of disabling button.
     * @param {ButtonDisableWhen} disableWhen Either a number of a page or a function that determines number of a page.
     * @returns {this} Button data.
     */
    _setDisableWhen(disableWhen) {
        if (typeof disableWhen === "number" && (!Number.isInteger(disableWhen) || disableWhen < -1))
            throw new RangeError("DisableWhen should return natural number or minus one to be always turned on.");
        this._disableWhen = disableWhen;
        return this;
    }
    ;
    /**
     * Setups class from style or ButtonData.
     * @param {ButtonStyling | ButtonAction} data Style or data.
     * @returns {void}
     */
    _setup(data) {
        if (!(data instanceof ButtonData)) {
            this.setStyle(data);
        }
        else if (data.style) {
            this._style = data.style;
        }
        ;
        return;
    }
    ;
}
;
exports.default = ButtonData;
